File: 1416-restore-the-array/1416-restore-the-array.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var numberOfArrays = function(s, k) {
  const n = s.length;
  const mod = 1e9 + 7;
  const dp = new Array(n + 1).fill(0);
  dp[0] = 1;
  
  for (let i = 0; i < n; i++) {
    if (s[i] === '0') continue;
    
    let num = 0;
    for (let j = i; j < n; j++) {
      num = num * 10 + +s[j];
      if (num > k) break;
      
      dp[j + 1] = (dp[j + 1] + dp[i]) % mod;
    }
  }
  
  return dp[n];
};

File: 1416-restore-the-array/1416-restore-the-array.py
class Solution:
    def numberOfArrays(self, s: str, k: int) -> int:
        @cache
        def dfs(i):
            if i == len(s):
                return 1
            if s[i] == '0':
                return 0
            ans, num = 0, ''
            for j in range(i, len(s)):
                num += s[j]
                if int(num) > k:
                    break
                ans += dfs(j + 1)
            return ans % (10**9 + 7)
          
        return dfs(0)